source/mbed_trace.c

```diff
@@ -267,7 +290,10 @@ void mbed_tracef(uint8_t dlevel, const char *grp, const char *fmt, ...)
 }
 void mbed_vtracef(uint8_t dlevel, const char* grp, const char *fmt, va_list ap)
 {
+    MUTEX_WAIT(); //MUTEX_RELEASE() must be called before returning
+
     if (NULL == m_trace.line) {
+        MUTEX_RELEASE();
```

Comment:
In Nanostack, we use lots of `goto` statements. They have a place for these kind of error handling.

```
void f(...)
{
    MUTEX_WAIT();

    if (...) { // There might be multiple cases where we need to return.
        goto end;
    }
    ...rest of the function...
end:
    MUTEX_RELEASE();
}
```

In Nanostack, those case are usually `drop:` to mean that we dropped this packet, and there might be multiple reasons to do so, but only one way to drop/free it.